library/src/main/scala/scala/scalajs/js/Union.scala

```diff
@@ -0,0 +1,65 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package scala.scalajs.js
+
+import scala.language.implicitConversions
+
+/** Value of type A or B (union type).
+ *
+ *  Scala does not have union types, but they are important to many
+ *  interoperability scenarios. This type provides a (partial) encoding of
+ *  union types using implicit evidences.
+ */
+@scala.scalajs.js.annotation.RawJSType // Don't do this at home!
+sealed trait |[A, B] // scalastyle:ignore
+
+object | { // scalastyle:ignore
+  /** Evidence that `A <: B`, taking top-level `|`-types into account. */
+  sealed trait Evidence[-A, +B]
+
+  /** A unique (and typically dead-code-eliminated away) instance of
+   *  `Evidence`.
+   */
+  private object ReusableEvidence extends Evidence[scala.Any, scala.Any]
+
+  abstract sealed class EvidenceLowestPrioImplicits {
+    /** If `A <: B2`, then `A <: B1 | B2`. */
+    implicit def right[A, B1, B2](implicit ev: Evidence[A, B2]): Evidence[A, B1 | B2] =
+      ReusableEvidence.asInstanceOf[Evidence[A, B1 | B2]]
+  }
+
+  abstract sealed class EvidenceLowPrioImplicits extends EvidenceLowestPrioImplicits {
+    /** `Int <: Double`, because that's true in Scala.js. */
+    implicit def intDouble: Evidence[Int, Double] =
+      ReusableEvidence.asInstanceOf[Evidence[Int, Double]]
```

Comment:
If you come from a `x: Byte` that you want to assign to an `Int | String`, for example, you can always force its conversion a priori with `x.toInt`, which will typecheck as an `Int | String`.

However, if you already have a `Byte | String`, then you cannot assign it to an `Int | String`. You have to cast. But then again, there are still plenty of cases that are not supported by this encoding, and you have to cast anyway. Basically every time that a cast of one of your base types is needed, you have to cast the whole thing instead. Or for example if you have to use a `js.use(x).as[T]` for one of your elements, you have the same problem. But all of that only happens when you already have an `A | B` and you want to upcast it to `A2 | B` for some conversion of `A` to `A2`. And as I explained in another comment, it is not possible to support that in general.

I initially wanted to *not* have `intDouble` either. But it was just too impractical a limitation for that particular case.

We could also have an additional namespace for "extra" evidences in `|.ExtraEvidences`. One could `import |.ExtraEvidences._` when needed. But well ... not super nice either, and it would still have limitations.